tests/restore_test.go

```diff
@@ -0,0 +1,683 @@
+		})
+	}
+
+	deleteWebhook := func(wh *admissionregistrationv1beta1.ValidatingWebhookConfiguration) {
+		waitDeleted(func() error {
+			return virtClient.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().Delete(wh.Name, &metav1.DeleteOptions{})
+		})
+	}
+
+	Context("With simple VM", func() {
+		var vm *v1.VirtualMachine
+
+		BeforeEach(func() {
+			var err error
+			vmiImage := cd.ContainerDiskFor(cd.ContainerDiskCirros)
+			vmi := tests.NewRandomVMIWithEphemeralDiskAndUserdata(vmiImage, "#!/bin/bash\necho 'hello'\n")
+			vm = tests.NewRandomVirtualMachine(vmi, false)
+			vm, err = virtClient.VirtualMachine(tests.NamespaceTestDefault).Create(vm)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			deleteVM(vm)
+		})
+
+		Context("and no snapshot", func() {
+			It("should reject restore", func() {
+				restore := createRestoreDef(vm, "foobar")
+
+				_, err := virtClient.VirtualMachineRestore(vm.Namespace).Create(restore)
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(ContainSubstring("VirtualMachineSnapshot \"foobar\" does not exist"))
+			})
+		})
+
+		Context("and good snapshot exists", func() {
+			var err error
+			var snapshot *snapshotv1.VirtualMachineSnapshot
+			var webhook *admissionregistrationv1beta1.ValidatingWebhookConfiguration
+
+			BeforeEach(func() {
+				snapshot = createSnapshot(vm)
+			})
+
+			AfterEach(func() {
+				deleteSnapshot(snapshot)
+				if webhook != nil {
+					deleteWebhook(webhook)
+				}
+			})
+
+			It("should successfully restore", func() {
+				var origSpec *v1.VirtualMachineSpec
+
+				Eventually(func() bool {
+					var updatedVM *v1.VirtualMachine
+					vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					origSpec = vm.Spec.DeepCopy()
+					Expect(origSpec.Template.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory]).To(Equal(resource.MustParse("64M")))
+
+					vm.Spec.Template.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("128M")
+					updatedVM, err = virtClient.VirtualMachine(vm.Namespace).Update(vm)
+					if errors.IsConflict(err) {
+						return false
+					}
+					vm = updatedVM
+					Expect(err).ToNot(HaveOccurred())
+					Expect(vm.Spec.Template.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory]).To(Equal(resource.MustParse("128M")))
+					return true
+				}, 180*time.Second, time.Second).Should(BeTrue())
+
+				restore := createRestoreDef(vm, snapshot.Name)
+
+				restore, err = virtClient.VirtualMachineRestore(vm.Namespace).Create(restore)
+				Expect(err).ToNot(HaveOccurred())
+
+				restore = waitRestoreComplete(restore, vm)
+				Expect(restore.Status.Restores).To(HaveLen(0))
+				Expect(restore.Status.DeletedDataVolumes).To(HaveLen(0))
+
+				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &metav1.GetOptions{})
+				Expect(vm.Spec).To(Equal(*origSpec))
```

Comment:
would it make sense to validate something written to disk is restored as well?

Reply:
covered here:  https://github.com/kubevirt/kubevirt/blob/98d5169f37b007a6c6317e183f75056976ae135d/tests/restore_test.go#L345-L683